### DISCUSSION TITLE
Is it possible to set a log4j2 logger for http4s ?

### DISCUSSION THREAD

**Discussion by yuriy-yarosh** [http4s/http4s#4041] [Q&A] [answered]
I'm with struggling setting log4j2 logger for http4s...

Here's my deps
```
  lazy val Log4jVersion = "2.14.0"
  
  lazy val logging =
       "org.apache.logging.log4j" % "log4j-api" % Log4jVersion ::
       "org.apache.logging.log4j" %% "log4j-api-scala" % "12.0" ::
       "org.apache.logging.log4j" % "log4j-core" % Log4jVersion ::
       "org.apache.logging.log4j" % "log4j-slf4j-impl" % Log4jVersion :: Nil
``` 

It looks like log4s picks log4j correctly, but...

log4j simply doesn't log much out of the box, and I was unable to set the default logging level.
Although the log4j2 logger itself does work but shows only error messages.
Log4J2 is a bit picky with it's configs when they're in different ClassPaths, and I don't really know how to deal with that in Scala.

Could you guys, please, provide a small log4j2 example and improve docs accordingly ?

Thank you.

**Comment by Daenyth** [comment]
Off hand, I'm fairly sure that http4s uses slf4j under the hood, so that *should* work.

Can you show how your routes/client is set up? Are you using applying the default logging middleware or do you have a custom one?

**Reply by yuriy-yarosh** [reply]
> http4s uses slf4j under the hood

Yes, it does.

>  so that should work

It does work but log4j2 doesn't pickup config from unmanagedResources, because it requires for config to be in the same ClassPath, which is quite difficult in case of scala and sbt.

> Can you show how your routes/client is set up?

Sure.

```scala
val httpApp = Router(
      "/" -> Home.service,
      "/auth" -> Auth.service,
      "/devices" -> Devices.service,
      "/org" -> Organization.service,
      "/stats" -> Stats.service,
      "/suite" -> Suite.service,
      "/user" -> User.service
    ).orNotFound

IO
      .fromEither(ConfigSource.defaultApplication.load[AppConfig].toEither)
      .flatMap(conf => {
        BlazeServerBuilder[IO](ExecutionContext.global)
          .bindHttp(conf.port.value, conf.host.value)
          .withHttpApp(httpApp)
          .withNio2(true)
          .withSocketReuseAddress(true)
          .resource
          .use(_ => IO.never)
          .as(ExitCode.Success)
      })
```

My sbt config

```scala
lazy val app = (project in file("."))
  .settings(commonSettings: _*)
  .settings(Dependencies.compilerPlugins.map(addCompilerPlugin): _*)
  .settings(
    name := "app",
    assemblyJarName in assembly := "app.jar",
    libraryDependencies ++= Dependencies.all,
    Compile / scalaSource := baseDirectory.value / "server" / "src",
    Compile / unmanagedResourceDirectories ++= baseDirectory.value / "app" / "build" ::
      baseDirectory.value / "server" / "resources" ::
      Nil,
    Test / scalaSource := baseDirectory.value / "server" / "test",
    mainClass in (Compile, run) := Some(mainClassName),
    mainClass in reStart := Some(mainClassName),
    bloopMainClass in (Compile, run) := Some(mainClassName),
    mainClass in assembly := Some(mainClassName)
  )
```

> Are you using applying the default logging middleware or do you have a custom one?

Nope, everything stock.
I'm getting INFO level notifications when I'm using sbt, but it disappears in app itself and in the build.

I'd like to get the same INFO level output in prod build, but was unable to configure it properly for some reason.

I might create another issue in `log4j-api-scala` because it's really weird.

@Daenyth sorry for bothering.

**Comment by yuriy-yarosh** [comment, marked as answer]
Figured it out.

log4j2 does need jackson yaml deps to be able to actually read.
It's pretty frustrating why it skips a yaml config file and doesn't throw an exception (or log something) when the jackson dependency is missing (best exception handling ever).

So, it worked with

```scala
  lazy val logging = "org.apache.logging.log4j" % "log4j-api"               % Log4jVersion ::
    "org.apache.logging.log4j"                 %% "log4j-api-scala"         % "12.0" ::
    "org.apache.logging.log4j"                  % "log4j-core"              % Log4jVersion ::
    "org.apache.logging.log4j"                  % "log4j-slf4j-impl"        % Log4jVersion ::
    "org.slf4j"                                 % "slf4j-api"               % "1.7.30" ::
    "com.fasterxml.jackson.core"                % "jackson-core"            % Jacksonversion ::
    "com.fasterxml.jackson.dataformat"          % "jackson-dataformat-yaml" % Jacksonversion ::
    Nil
```
log4j2 deps

Well, had to figure it out by feeding a yml config manually, and default config kicks in only after Config is being .reconfigured(). 
Hopefully someone will find it helpful.
